### ISSUE TITLE
[MacOS] It is currently impossible to use pyenv to install/build python that links against OpenSSL 1.1

### ISSUE THREAD

**Issue by cculianu** [pyenv/pyenv#1268]
### Description
- [x] Platform information: MacOSX 10.13 High Sierra
- [x] OS architecture: amd64
- [x] pyenv version: 1.2.9-2-g6309aaf
- [x] Python version: Any/All
- [x] C Compiler information: clang-1000.11.45.5 
- [ ] Please attach verbose build log as gist

---

Hi all,

I'm on OSX with Brew.

It's currently impossible to get pyenv to build & install a version of Python linked against OpenSSL 1.1.  Passing the arg "--with-openssl=" to Python via PYTHON_CONFIGURE_OPTS or CONFIGURE_OPTS doesn't have the intended effect as pyenv also inserts its own path to openssl (1.0). Any combination of this plus LDFLAGS + CFLAGS also doesn't take.  They show up on the command-line but the pyenv-specified flags are the ones that end up taking (the pyenv flags point to old openssl 1.0).

If I unistall openssl 1.0 and just leave 1.1 installed, pyenv insists on downloading OpenSSL 1.0 and compiling it privately.

The only way to do it is to "fool" pyenv and the python build process by uninstalling openssl1.0 and symlinking /usr/local/opt/openssl to /usr/local/opt/openssl@1.1. (While this "works", this is a hack and a nightmare for those of use that work on open source projects where our users expect to be able to reproducibly and easily deterministically build an app.)

The situation is pretty frustrating because OpenSSL 1.1 offers newer features that end up showing up in the Python SSL API's.  Here is 1 exampe: the `scrypt` function which is not available unless Python is linked against OpenSSL 1.1: https://docs.python.org/3/library/hashlib.html#hashlib.scrypt

**Suggestions**: Allow users to override the openssl options and/or have pyenv understand that OpenSSL1.1 is a thing and allow users to select it at build/install-time.

Not having access to OpenSSL 1.1 from within Python is very limiting!

Thanks in advance.

**Comment by cculianu** [comment]
Update: One of the places where a patch could be added would be here:

https://github.com/pyenv/pyenv/blob/6309aaf295b05fece4abb52ba1000f98e40390e7/plugins/python-build/bin/python-build#L1494

Either it could:

1. Auto-detect openssl1.1 if it exists and prefer that or
2. Allow users to tell the build script to look for openssl1.1 with say some option like --use-openssl1.1 or something.

Let me know if this approach sounds reasonable and if you guys are open to a PR and maybe I can put together a PR to allow this to happen!

**Comment by yyuu** [comment]
At least I believe that python-build should pick up openssl built by homebrew on macOS by default. For now I have no idea if there is some difference beteeen openssl 1.0 and 1.1. I may gonna try reproducing it later. (IIRC "--with-openssl" might work with 3.7+ only anf it might be the cause of the issue 🤔 

https://github.com/pyenv/pyenv/blob/master/plugins/python-build/bin/python-build#L1494

**Comment by cculianu** [comment]
Thanks for the quick reply!

@yyuu I tried 3.7.2, 3.7.1, 3.7.0, all the 3.6's.  It appears we are commenting at the same time.  The issue is here: https://github.com/pyenv/pyenv/blob/6309aaf295b05fece4abb52ba1000f98e40390e7/plugins/python-build/bin/python-build#L1494

^^ that code only looks for openssl 1.0.

And yes.. there is a difference between 1.1 -- 1.1 offers more hashing and other functions that Python knows about can can use.. :)

**Comment by cculianu** [comment]
PS: If you wanted to ask brew for the prefix to openssl1.1 it would be:

`brew --prefix openssl@1.1`

Brew allows both libs to be installed side-by-side and simple `openssl` is just 1.0.

It would actually be super-nice if the python-build plugin detected whether a user specified an override: --with-openssl= and if so, just respected the user's override and used that as the prefix, rather than overriding the override. :)

**Comment by yyuu** [comment]
jfyi, I'm using pyenv with OpenSSL 1.1 on my Debian GNU/Linux box. As you can find in the python-build, this issue is macOS specific.

```
% ldd "$(pyenv root)/versions/3.7.0/lib/python3.7/lib-dynload/_ssl.cpython-37m-x86_64-linux-gnu.so"
        linux-vdso.so.1 (0x00007fff133fa000)
        libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007fa0e5198000)
        libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007fa0e4eb0000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa0e4e8f000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa0e4cce000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fa0e4cc9000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fa0e5260000)
```

**Comment by cculianu** [comment]
Ah. Indeed it is.

**Comment by chrahunt** [comment]
I think this can be considered fixed given the PR that fixed #839 has been merged. @cculianu can you confirm the behavior works for you?

**Comment by cculianu** [comment]
Oh yes definitely works.  Sorry I forgot to close it -- I figured @yyuu  would close.

I've been using it and it works like a dream.  I use it daily.
